File: src/auth_test.rs
use super::*;

#[test]
fn test_access_policy_public_mode() {
    let policy = AccessPolicy { expected_secret: None };
    assert_eq!(policy.check_access(""), Ok(()));
    assert_eq!(policy.check_access("any-secret"), Ok(()));
}

#[test]
fn test_access_policy_private_mode() {
    let policy = AccessPolicy { expected_secret: Some("secret123".into()) };

    // Correct secret
    assert_eq!(policy.check_access("secret123"), Ok(()));

    // Incorrect secrets
    assert_eq!(policy.check_access("wrong"), Err(AuthError::Forbidden));
    assert_eq!(policy.check_access(""), Err(AuthError::Forbidden));
    assert_eq!(policy.check_access("secret1234"), Err(AuthError::Forbidden));
}

#[test]
fn test_constant_time_eq() {
    assert!(constant_time_eq("abc", "abc"));
    assert!(!constant_time_eq("abc", "abd"));
    assert!(!constant_time_eq("abc", "abcd"));
    assert!(constant_time_eq("", ""));
}

#[test]
fn test_constant_time_eq_different_lengths_short() {
    assert!(!constant_time_eq("a", "ab"));
    assert!(!constant_time_eq("ab", "a"));
}

#[test]
fn test_constant_time_eq_all_same_bytes() {
    assert!(constant_time_eq("aaa", "aaa"));
    assert!(!constant_time_eq("aaa", "baa"));
}

#[test]
fn test_access_policy_from_env_treated_as_none() {
    let policy = AccessPolicy { expected_secret: Some("".into()) };
    assert_eq!(policy.check_access(""), Ok(()));
    assert_eq!(policy.check_access("any"), Err(AuthError::Forbidden));
}

#[test]
fn test_access_policy_empty_string_as_none() {
    let policy = AccessPolicy { expected_secret: None };
    assert_eq!(policy.check_access(""), Ok(()));
    assert_eq!(policy.check_access("anything"), Ok(()));
}

#[test]
fn test_access_policy_private_mode_case_sensitive() {
    let policy = AccessPolicy { expected_secret: Some("Secret123".into()) };
    assert_eq!(policy.check_access("Secret123"), Ok(()));
    assert_eq!(policy.check_access("secret123"), Err(AuthError::Forbidden));
    assert_eq!(policy.check_access("SECRET123"), Err(AuthError::Forbidden));
}

#[test]
fn test_access_policy_private_mode_similar() {
    let policy = AccessPolicy { expected_secret: Some("test-secret-123".into()) };
    assert_eq!(policy.check_access("test-secret-123"), Ok(()));
    assert_eq!(policy.check_access("test-secret-124"), Err(AuthError::Forbidden));
    assert_eq!(policy.check_access("test-secret-12"), Err(AuthError::Forbidden));
    assert_eq!(policy.check_access("test-secret123"), Err(AuthError::Forbidden));
}

#[test]
fn test_constant_time_eq_long_strings() {
    let a = "this is a much longer test string for testing";
    let b = "this is a much longer test string for testing";
    let c = "this is a much longer test string for testing!";
    assert!(constant_time_eq(a, b));
    assert!(!constant_time_eq(a, c));
}

#[test]
fn test_constant_time_eq_unicode() {
    assert!(constant_time_eq("🦀", "🦀"));
    assert!(!constant_time_eq("🦀", "🐡"));
    assert!(constant_time_eq("こんにちは", "こんにちは"));
    assert!(!constant_time_eq("こんにちは", "こんばんは"));
}

#[test]
fn test_access_policy_new_some() {
    let policy = AccessPolicy::new(Some("mypass".into()));
    assert_eq!(policy.check_access("mypass"), Ok(()));
    assert_eq!(policy.check_access("wrong"), Err(AuthError::Forbidden));
}

#[test]
fn test_access_policy_new_none() {
    let policy = AccessPolicy::new(None);
    assert_eq!(policy.check_access("anything"), Ok(()));
}

#[test]
fn test_access_policy_new_empty_string() {
    let policy = AccessPolicy::new(Some("".into()));
    assert_eq!(policy.check_access("anything"), Ok(()));
}

File: src/cloudflare/auth.test.ts
import { describe, it, expect } from 'vitest';
import { AuthError, AccessPolicy } from './auth.ts';

describe('AccessPolicy', () => {
  it('public mode allows any secret', () => {
    const policy = new AccessPolicy(null);
    expect(() => { policy.checkAccess(''); }).not.toThrow();
    expect(() => { policy.checkAccess('any-secret'); }).not.toThrow();
  });

  it('private mode rejects wrong secret', () => {
    const policy = new AccessPolicy('secret123');
    expect(() => { policy.checkAccess('secret123'); }).not.toThrow();
    expect(() => { policy.checkAccess('wrong'); }).toThrow(AuthError);
    expect(() => { policy.checkAccess(''); }).toThrow(AuthError);
    expect(() => { policy.checkAccess('secret1234'); }).toThrow(AuthError);
  });

  it('constant-time comparison', () => {
    // Let's test the underlying impl via policy behavior
    const policy = new AccessPolicy('abc');
    expect(() => { policy.checkAccess('abc'); }).not.toThrow();
    expect(() => { policy.checkAccess('abd'); }).toThrow(AuthError);
    expect(() => { policy.checkAccess('abcd'); }).toThrow(AuthError);
  });

  it('empty string via constructor becomes public mode', () => {
    const policy = new AccessPolicy('');
    expect(() => { policy.checkAccess(''); }).not.toThrow();
    expect(() => { policy.checkAccess('any'); }).not.toThrow();
  });

  it('null secret treated as public mode', () => {
    const policy = new AccessPolicy(null);
    expect(() => { policy.checkAccess(''); }).not.toThrow();
    expect(() => { policy.checkAccess('anything'); }).not.toThrow();
  });

  it('private mode is case sensitive', () => {
    const policy = new AccessPolicy('Secret123');
    expect(() => { policy.checkAccess('Secret123'); }).not.toThrow();
    expect(() => { policy.checkAccess('secret123'); }).toThrow(AuthError);
    expect(() => { policy.checkAccess('SECRET123'); }).toThrow(AuthError);
  });

  it('private mode with similar secrets', () => {
    const policy = new AccessPolicy('test-secret-123');
    expect(() => { policy.checkAccess('test-secret-123'); }).not.toThrow();
    expect(() => { policy.checkAccess('test-secret-124'); }).toThrow(AuthError);
    expect(() => { policy.checkAccess('test-secret-12'); }).toThrow(AuthError);
    expect(() => { policy.checkAccess('test-secret123'); }).toThrow(AuthError);
  });

  it('constructor with Some secret', () => {
    const policy = new AccessPolicy('mypass');
    expect(() => { policy.checkAccess('mypass'); }).not.toThrow();
    expect(() => { policy.checkAccess('wrong'); }).toThrow(AuthError);
  });

  it('constructor with None', () => {
    const policy = new AccessPolicy(null);
    expect(() => { policy.checkAccess('anything'); }).not.toThrow();
  });

  it('constructor with empty string becomes public', () => {
    const policy = new AccessPolicy('');
    expect(() => { policy.checkAccess('anything'); }).not.toThrow();
  });
});
